### PR TITLE
Add exporting of a samples table to Terra

### PR DESCRIPTION
**Pull request by bfcrampton**
FYI @malathi13

### CODE REVIEW

**Comment by bfcrampton** [comment]
screen shot 2018-10-04 at 3 08 20 pm

File: api/data_explorer/controllers/facets_controller.py [hunk 1]
```diff
@@ -102,65 +112,3 @@ def facets_get(filter=None, extraFacets=None):  # noqa: E501
 
     return FacetsResponse(
         facets=facets, count=es_response._faceted_search.count())
-
-
-def deserialize(filter_arr, es_facets):
```
Comment (malathi13) [inline api/data_explorer/controllers/facets_controller.py]: Why is this moved to util? Is this method used somewhere else?

File: api/data_explorer/controllers/facets_controller.py [hunk 2]
```diff
@@ -13,17 +12,6 @@
 from data_explorer.util.dataset_faceted_search import DatasetFacetedSearch
 
 
-def _is_histogram_facet(facet):
```
Comment (malathi13) [inline api/data_explorer/controllers/facets_controller.py]: Same here, is this used somewhere apart from facets_controller?

File: api/data_explorer/controllers/facets_controller.py (hunk 1 above)
Comment (bfcrampton) [reply api/data_explorer/controllers/facets_controller.py]: Yeah, in `export_url_controller`.

File: api/data_explorer/controllers/facets_controller.py (hunk 2 above)
Comment (bfcrampton) [reply api/data_explorer/controllers/facets_controller.py]: Used in `deserialize` which is used by `export_url_controller`.

File: util/dump_index.py [hunk 3]
```diff
@@ -9,7 +9,7 @@
 - Run:
 
 virtualenv ~/virtualenv/dump_index
-source ~/virtualenv/dump_index/bin/activate
+source ~/virtualenv/elasticsearch/bin/dump_index
```
Comment (malathi13) [inline util/dump_index.py]: Why has this changed?

File: api/data_explorer/controllers/export_url_controller.py [hunk 4]
```diff
@@ -103,6 +107,22 @@ def _get_entities_dict(cohort_name, query):
                 'query': query
             }
         })
+
+    for doc in doc_generator:
+        for sample in doc.get('samples', []):
+            sample_id = sample['sample_id']
+            export_sample = {
+                k.replace('.', '_'): v
```
Comment (melissachang) [inline api/data_explorer/controllers/export_url_controller.py]:
```
# TODO: Remove replace after new Entity Service is in prod.
```

Comment (melissachang) [inline api/data_explorer/controllers/export_url_controller.py]: Let's use "Has_phase1_chrMT_SNPs" instead of "verily-public-data_human_genome_variants_1000_genomes_sample_info_Has_phase1_chrMT_SNPs" for column name

File: api/data_explorer/controllers/export_url_controller.py [hunk 5]
```diff
@@ -259,15 +279,34 @@ def _get_filter_query(filters):
     return query
 
 
+def _get_doc_generator(filter_arr):
```
Comment (melissachang) [inline api/data_explorer/controllers/export_url_controller.py]: I think we should query BQ instead of ES. The code will be easier to understand, and also the columns would be in the right order. (Right now columns are in random order.) Note this code will go away once Saturn UI can run SQL queries.

I have some ideas for addressing slowness:
- If user selected cohort, kick off creating GCS file when they click export button, instead of when they save cohort name dialog
- Could build JSON file representing everything on page load

File: api/data_explorer/controllers/export_url_controller.py [hunk 6]
```diff
@@ -9,14 +9,18 @@
 import time
 import urllib
 
+from collections import OrderedDict
```
Comment (melissachang) [inline api/data_explorer/controllers/export_url_controller.py]: I simulated exporting no cohort by checking all checkboxes in one facet.

Bad Gateway
The proxy server received an invalid response from an upstream server.
Apache Server at rawls.dsde-prod.broadinstitute.org Port 443

Shouldn't block this PR, but ideally we should fix this at some point

Comment (melissachang) [inline api/data_explorer/controllers/export_url_controller.py]: Export with no cohort doesn't work

Comment (melissachang) [inline api/data_explorer/controllers/export_url_controller.py]: I no longer get "Data too long for column" after taking out fastq, so let's do that. You're a Project Owner on verily-public-data, so I'll let you handle the switchover.

It might make sense to send/merge the 2 PRs making the switch before this PR.

**Comment by bfcrampton** [comment]
screen shot 2018-10-09 at 2 42 21 pm

File: api/data_explorer/controllers/export_url_controller.py (hunk 4 above)
Comment (bfcrampton) [reply api/data_explorer/controllers/export_url_controller.py]: No longer relevant (done in indexer and we're not including dataset/table names).

File: api/data_explorer/controllers/export_url_controller.py [hunk 7]
```diff
@@ -259,15 +279,34 @@ def _get_filter_query(filters):
...
-    entities = _get_entities_dict(cohortname, query)
+
+    # TODO(malathir): Add extraFacets to this endpoint
+    query = _get_filter_query(filter_arr)
+    cohortname = data['cohortName'].replace(" ", "_")
+    doc_generator = _get_doc_generator(filter_arr)
+
+    entities = _get_entities_dict(cohortname, query, doc_generator)
     current_app.logger.info('Entity JSON: %s' % json.dumps(entities))
+
```
[the remark this answers is not in the document]
Comment (bfcrampton) [reply api/data_explorer/controllers/export_url_controller.py]: Done.

File: api/data_explorer/controllers/export_url_controller.py (hunk 4 above)
Comment (bfcrampton) [reply api/data_explorer/controllers/export_url_controller.py]: Done (indexer PR).

File: api/data_explorer/controllers/export_url_controller.py (hunk 5 above)
Comment (bfcrampton) [reply api/data_explorer/controllers/export_url_controller.py]: Per offline discussion, this would likely be quite slow. However this has all been moved to the indexer, so doesn't really matter.

File: api/data_explorer/controllers/export_url_controller.py (hunk 6 above)
Comment (bfcrampton) [reply api/data_explorer/controllers/export_url_controller.py]: Removed fastq.

File: api/data_explorer/controllers/export_url_controller.py [hunk 8]
```diff
@@ -103,6 +107,22 @@ def _get_entities_dict(cohort_name, query):
...
+        for sample in doc.get('samples', []):
+            sample_id = sample['sample_id']
+            export_sample = {
+                k.replace('.', '_'): v
+                for k, v in sample.iteritems()
+                if k != 'sample_id' and not k.startswith('_has_')
+            }
+
+            entities.append({
+                'entityType': 'samples',
```
[the remark this answers is not in the document]
Comment (bfcrampton) [reply api/data_explorer/controllers/export_url_controller.py]: Done.

File: api/data_explorer/controllers/export_url_controller.py (hunk 5 above)
Comment (bfcrampton) [reply api/data_explorer/controllers/export_url_controller.py]: Sorry that's not totally true, we still query ES to get the sample_set.

File: api/data_explorer/controllers/export_url_controller.py (hunk 6 above)
Comment (bfcrampton) [reply api/data_explorer/controllers/export_url_controller.py]: Fixed, now just exports the full samples table (no sample set).

Comment (bfcrampton) [reply api/data_explorer/controllers/export_url_controller.py]: Selecting all checkboxes works and correctly exports sample set containing all samples.

**Comment by bfcrampton** [comment]
Still working on fixing tests.

File: api/data_explorer/controllers/export_url_controller.py [hunk 9]
```diff
@@ -103,23 +107,62 @@ def _get_entities_dict(cohort_name, query):
                 'query': query
             }
         })
+
+    if cohort_name:
+        sample_items = []
```
Comment (melissachang) [inline api/data_explorer/controllers/export_url_controller.py]: I think it's cleaner if you call _get_doc_generator() here, instead of passing in doc_generator. (And move  _get_doc_generator() to before this method)

File: api/data_explorer/controllers/export_url_controller.py [hunk 10]
```diff
@@ -259,15 +302,41 @@ def _get_filter_query(filters):
...
-    cohortname = data['cohortName']
-    cohortname = cohortname.replace(" ", "_")
-    entities = _get_entities_dict(cohortname, query)
-    current_app.logger.info('Entity JSON: %s' % json.dumps(entities))
+
+    # TODO(malathir): Add extraFacets to this endpoint
+    query = _get_filter_query(filter_arr)
+    cohort_name = data['cohortName'].replace(" ", "_")
+
+    # If we have not selected a cohort there's no need accumulate
```
Comment (melissachang) [inline api/data_explorer/controllers/export_url_controller.py]:
```
# If a cohort was selected, get Elasticsearch documents for the cohort so we can create sample_set entity.
```

File: api/data_explorer/controllers/export_url_controller.py [hunk 11]
```diff
@@ -103,23 +107,62 @@ def _get_entities_dict(cohort_name, query):
                 'query': query
             }
         })
+
+    if cohort_name:
```
Comment (melissachang) [inline api/data_explorer/controllers/export_url_controller.py]: Line 102 should be `if cohort_name`. There should be one `if cohort_name` that includes both cohort and sample_set.

File: api/data_explorer/util/elasticsearch_util.py [hunk 12]
```diff
@@ -153,12 +102,95 @@ def _get_bucket_interval(field_range):
...
+        s = s.replace(char, '_')
+    s = s.lower()
+    # Remove leading underscore.
+    if s.find('_', 0, 1) == 0:
+        s = s.lstrip('_')
+    print('Index name: %s' % s)
+    return s
+
+
+def deserialize(filter_arr, es_facets):
```
Comment (melissachang) [inline api/data_explorer/util/elasticsearch_util.py]: Since this is moving out of facets_controller, it should have a more descriptive name. `get_facet_value_dict`

File: util/dump_index.py (hunk 3 above)
Comment (melissachang) [inline util/dump_index.py]: Revert, the original is correct

File: api/data_explorer/__main__.py [hunk 13]
```diff
@@ -226,6 +226,7 @@ def _process_export_url():
     app.app.config['AUTHORIZATION_DOMAIN'] = ''
     app.app.config['DEPLOY_PROJECT_ID'] = ''
     app.app.config['EXPORT_URL_GCS_BUCKET'] = ''
+    app.app.config['EXPORT_SAMPLES_URL_GCS_BUCKET'] = ''
```
[the remark this answers is not in the document]
Comment (bfcrampton) [reply api/data_explorer/__main__.py]: Done.

File: api/data_explorer/controllers/export_url_controller.py [hunk 14]
```diff
@@ -9,14 +9,18 @@
 import time
 import urllib
 
+from collections import OrderedDict
+from data_explorer.models.export_url_response import ExportUrlResponse  # noqa: E501
```
[the remark this answers is not in the document]
Comment (bfcrampton) [reply api/data_explorer/controllers/export_url_controller.py]: Done.

File: api/data_explorer/controllers/export_url_controller.py (hunk 11 above)
Comment (bfcrampton) [reply api/data_explorer/controllers/export_url_controller.py]: Done.

File: api/data_explorer/controllers/export_url_controller.py [hunk 15]
```diff
@@ -103,23 +107,56 @@ def _get_entities_dict(cohort_name, query):
...
+    # Remove the leading '[' character since this is being concatenated with the
+    # sample entities JSON, which has the trailing ']' stripped in the indexer.
+    entities_json = ',%s' % entities_json[1:]
+    blob.upload_from_string(entities_json)
+
+    merged = export_bucket.blob(_random_str())
+    merged.upload_from_string('')
+    merged.compose([samples_blob, blob])
+
+    current_app.logger.info('Wrote %s' % merged.path)
```
[the remark this answers is not in the document]
Comment (bfcrampton) [reply api/data_explorer/controllers/export_url_controller.py]: Done.

File: api/data_explorer/controllers/export_url_controller.py (hunk 10 above)
Comment (bfcrampton) [reply api/data_explorer/controllers/export_url_controller.py]: Call is gone since I moved the call to get_doc_generator into _get_entities_dict, but added similar comment up there.

File: api/data_explorer/util/elasticsearch_util.py (hunk 12 above)
Comment (bfcrampton) [reply api/data_explorer/util/elasticsearch_util.py]: Done.

File: util/dump_index.py (hunk 3 above)
Comment (bfcrampton) [reply util/dump_index.py]: Ugh not sure why this keeps reverting, done.

File: api/data_explorer/controllers/export_url_controller.py (hunk 9 above)
Comment (bfcrampton) [reply api/data_explorer/controllers/export_url_controller.py]: Oops did this.
